File: problems/tier1-foundation/006-discount-engine/boilerplate/java/part2/Guided.java
import java.util.*;

// Data class (given).
class CartItem {
    public String name;
    public double price;
    public int quantity;
    public String category;

    public CartItem(String name, double price, int quantity, String category) {
        this.name = name;
        this.price = price;
        this.quantity = quantity;
        this.category = category;
    }

    public CartItem(String name, double price, int quantity) {
        this(name, price, quantity, "");
    }
}

// HINT: introduce an abstraction so new variants don't change existing code.
public class Solution {
    // HINT: start from what this must return, then work backwards to the state it needs.
    public static double apply_percentage_discount(List<CartItem> cart, double percentage) {
        // TODO: write your solution
        return 0.0;
    }

    // HINT: start from what this must return, then work backwards to the state it needs.
    public static double apply_flat_discount(List<CartItem> cart, double amount) {
        // TODO: write your solution
        return 0.0;
    }

    // HINT: start from what this must return, then work backwards to the state it needs.
    public static double apply_bogo(List<CartItem> cart, int buyCount, int freeCount) {
        // TODO: write your solution
        return 0.0;
    }

}

File: problems/tier1-foundation/006-discount-engine/boilerplate/python/part3/guided.py
# Data class (given).
class CartItem:
    def __init__(self, name, price, quantity, category=""):
        self.name = name
        self.price = price
        self.quantity = quantity
        self.category = category

# HINT: introduce an abstraction so new variants don't change existing code.

# HINT: start from what this must return, then work backwards to the state it needs.
def apply_percentage_discount(cart, percentage):
    # TODO: write your solution
    return None

# HINT: start from what this must return, then work backwards to the state it needs.
def apply_flat_discount(cart, amount):
    # TODO: write your solution
    return None

# HINT: start from what this must return, then work backwards to the state it needs.
def apply_bogo(cart, buyCount, freeCount):
    # TODO: write your solution
    return None

# HINT: start from what this must return, then work backwards to the state it needs.
def apply_percentage_with_eligibility(cart, percentage, minCartValue, requireFirstTimeUser, isFirstTimeUser, eligibleCategory):
    # TODO: write your solution
    return None
